pkg/bloomgateway/bloomgateway.go

```diff
@@ -355,19 +312,250 @@ func (g *Gateway) FilterChunkRefs(ctx context.Context, req *logproto.FilterChunk
 	})
 
 	response := make([]*logproto.GroupedChunkRefs, 0, len(req.Refs))
+	responseCount := 0
 	for {
 		select {
 		case <-ctx.Done():
-			return nil, ctx.Err()
+			return nil, errors.Wrap(ctx.Err(), "waiting for results")
 		case err := <-errCh:
-			return nil, err
+			return nil, errors.Wrap(err, "waiting for results")
 		case res := <-resCh:
-			level.Info(g.logger).Log("msg", "got result", "task", task.ID, "tenant", tenantID, "res", res)
+			responseCount++
+			// log line is helpful for debugging tests
+			// level.Debug(g.logger).Log("msg", "got partial result", "task", task.ID, "tenant", tenantID, "fp", res.Fp, "chunks", res.Chks.Len(), "progress", fmt.Sprintf("%d/%d", responseCount, len(req.Refs)))
+			if res.Chks.Len() > 0 {
+				response = append(response, &logproto.GroupedChunkRefs{
+					Tenant:      tenantID,
+					Fingerprint: uint64(res.Fp),
+					Refs:        convertToShortRefs(res.Chks),
+				})
+			}
 			// wait for all parts of the full response
-			response = append(response, res)
-			if len(response) == len(req.Refs) {
+			if responseCount == len(req.Refs) {
 				return &logproto.FilterChunkRefResponse{ChunkRefs: response}, nil
 			}
 		}
 	}
 }
+
+type workerConfig struct {
+	maxWaitTime time.Duration
+	maxItems    int
+}
+
+type workerMetrics struct {
```

Comment:
Since this file is getting quite big, I think it may make sense to create a new `metrics.go` file and move these there.

Same for the `metrics` struct here. In fact, I think in metrics.go we should define `metrics` to contain the worker and queue metrics:
```go
type metrics struct {
 workerMetrics
 queue.Metrics
 otherMetric prometheus.Summary
}
```